dot/telemetry/afg_authority_set.go

```diff
@@ -0,0 +1,44 @@
+// Copyright 2021 ChainSafe Systems (ON) Corp.
+// This file is part of gossamer.
+//
+// The gossamer library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The gossamer library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the gossamer library. If not, see <http://www.gnu.org/licenses/>.
+
+package telemetry
+
+// afgAuthoritySetTM is a telemetry message of type `afg.authority_set` which is
+// meant to be sent when authority set changes (generally when a round is initiated)
+type afgAuthoritySetTM struct {
+	// finalized hash and finalized number are sent by substrate, but not read by
+	// substrate telemetry
+	// FinalizedHash   *common.Hash `json:"hash"`
+	// FinalizedNumber string       `json:"number"`
+	AuthorityID    string `json:"authority_id"`
+	AuthoritySetID string `json:"authority_set_id"`
+	// Substrate creates an array of string of authority IDs. It JSON-serialises
+	// that array and send that as a string.
+	Authorities string `json:"authorities"`
+}
+
+// NewAfgAuthoritySetTM creates a new afgAuthoritySetTM struct.
+func NewAfgAuthoritySetTM(authorityID, authoritySetID string, authorities string) Message {
```

Comment:
You should return `*afgAuthoritySetTM` instead of the interface.
Export `afgAuthoritySetTM` if it's used outside of this package.
The caller should use the `Message` interface (up to the caller).

Reply:
Wherever it is used `NewAfgAuthoritySetTM` is enough. There was not any need to use `afgAuthoritySetTM` directly.

If you think returning the struct instead of interface is a better coding style, please provide a reason. Then, I am happy to change it.

Reply:
I had constructors creating unexported types and returning interfaces to hide implementation details for pretty much years, and switched to always return exported types and accept interfaces everywhere about a year ago, and it really does help a bunch, especially for testability. 

There is a [bunch of articles](https://www.google.com/search?q=accept+interfaces+return+concrete+types+go) explaining this more in depth, such as [that one](https://bryanftan.medium.com/accept-interfaces-return-structs-in-go-d4cab29a301b).

The only exception I have so far is when a constructor returns one of multiple implementations, then obviously it should return an interface.

But up to you, I guess we already have several constructors doing this which is out of scope.

Reply:
I have created an issue for this, so that I can change all such instances, if needed.